### PR TITLE
docs: reorganize function reference

### LINKED ISSUE #659: Improve pkgdown configuration for reference index
From https://github.com/igraph/rigraph/pull/638#issuecomment-1400511776 cc @szhorvat @ntamas @krlmlr

### PR DESCRIPTION
**Pull request by maelle**
Fix #659 
Based on #660

### CODE REVIEW

File: man/roxygen/meta.R [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  rd_family_title = list(
+    adjacency = "Adjacency matrices",
+    attributes = "Vertex, edge and graph attributes",
+    bipartite = "Bipartite graphs",
+    centrality = "Centrality measures",
+    coloring = "Graph coloring",
+    community = "Community detection",
+    components = "Connected components",
+    epi = "Processes on graphs",
+    eulerian = "Graph cycles",
```
Comment (maelle) [inline man/roxygen/meta.R:11]: or should it remain "eulerian"?

File: man/roxygen/meta.R [hunk 2]
```diff
@@ -0,0 +1,17 @@
+list(
+  rd_family_title = list(
+    adjacency = "Adjacency matrices",
+    attributes = "Vertex, edge and graph attributes",
+    bipartite = "Bipartite graphs",
+    centrality = "Centrality measures",
+    coloring = "Graph coloring",
+    community = "Community detection",
+    components = "Connected components",
+    epi = "Processes on graphs",
```
Comment (maelle) [inline man/roxygen/meta.R:10]: or should it remain "epi"?

File: man/roxygen/meta.R (hunk 1 above)
Comment (szhorvat) [reply man/roxygen/meta.R:11]: No, the idea is that as igraph gets more cycle-related functions, they will all go into this category.

E.g. C/igraph 0.10 has cycle basis functions. Like here: https://igraph.org/c/html/latest/igraph-Cycles.html

So I would also rename the identifiers, not just the titles.  `eulerian` -> `cycles`, `epi` -> `processes`, etc.

File: man/roxygen/meta.R (hunk 2 above)
Comment (szhorvat) [reply man/roxygen/meta.R:10]: I would rename it `processes`, as non epidemiological functionality may go into this section later.

**Comment by maelle** [comment]
my next goal is to get read of the "other" catch-all section, then to order things better.

**Comment by maelle** [comment]
is assortativity a structural property?

**Comment by maelle** [comment]
I'm constantly writing family as "famility" now :joy_cat:

**Comment by szhorvat** [comment]
> is assortativity a structural property?

Yes, we can call it that, and I do suggest placing it into that section.

Frankly, all properties are "structural properties" and that section should eventually be broken up, once we get there. Unfortunately, at the moment that section is used as a bit of a dumping ground.

**Comment by maelle** [comment]
does permute() relate to isomorphism, automorphism, or none of the two?

**Comment by szhorvat** [comment]
It _could_ go there, as it's useful in that context, though it's not exclusive to it. For example, I use vertex permutation in a non-isomorphism-related context when I need to ensure a certain ordering for the rows/columns of an adjacency matrix.

Alternatively, we could have a section for functions that _modify_ graphs ("Graph modification" or similar) and collect functions such as `permute`, `contract`, `rewire`, `connect`, etc. there.

Either of these solutions would work.

File: _pkgdown.yml [hunk 3]
```diff
@@ -46,54 +70,50 @@ reference:
...
   - has_concept("latent position vector samplers")
+- subtitle: Other samplers
+- contents:
+  - sample_
+  - sample_correlated_gnp_pair
+  - sample_fitness
+  - sample_fitness_pl
+  - sample_forestfire
+  - sample_tree
+  - sample_correlated_gnp
```
Comment (szhorvat) [inline _pkgdown.yml:91]: These should all go under Random graph models.

**Comment by szhorvat** [comment]
I'm sorry, I forgot that we do already have "Functions for manipulating graph structure". It could go there. Technically speaking, `permute` does not change the graph structure, only the graph representation. So perhaps shorten the name of this section to "Functions for manipulating graphs"?

**Comment by szhorvat** [comment]
@ntamas I just realized that `ecount()` and `vcount()` are deprecated in favour of `gsize()` and `gorder()`. They appear in `zzz-deprecate.R`.

I don't like this very much. Do you agree that we should keep `ecount` and `vcount` as fully supported names?

**Comment by maelle** [comment]
Since you mention it, I think some work on consolidating deprecation workflows+stages might be nice, for instance "nexus" functions are not really deprecated? I'd recommend exploring https://lifecycle.r-lib.org/

**Comment by szhorvat** [comment]
Sorry, I see you have it at https://maelle.github.io/rigraph/

Here are some more suggestions:

 - Can you put a single function in more than one section? `[simplify_and_colorize()](https://maelle.github.io/rigraph/reference/simplify.html)` can be placed under isomorphism. I implemented this function specifically for dealing with isomorphism testing of non-simple graphs.
 - The "Simple" section can be merged with "Functions for manipulating graphs"
 - I see that now there are subsections. "Random graph models" should be a subsection of Construction then.
 - Do not separate isomorphism from automorphisms. They are the same thing. Do not use the title "Topology" (which is mathematically inaccurate), just use "Isomoprhism". 
 - Rename Flow to "Maximum flow and connectivity"
 - Centalization should come right after Centrality measures, as it's closely related.
 - `triangles()` is missing for some reason.
 - Under Structural properties, let's make a subsection on "Triangles and transitivity". This would include all functions with "triangle" or "transitivity" in their names.

It's starting to look pretty good!

**Comment by szhorvat** [comment]
The "Graph matching" subsection of Structural properties should be moved to the top level (or possibly to a subsection of Isomorphism).

[is_matching()](https://maelle.github.io/rigraph/reference/matching.html) [is_max_matching()](https://maelle.github.io/rigraph/reference/matching.html) [max_bipartite_match()](https://maelle.github.io/rigraph/reference/matching.html) should be referred to simply as "Matching" and not "Graph matching". These are different concepts.

**Comment by szhorvat** [comment]
Processes on graphs could have two subsections, "Compartmental models" for SIR and "Random walks". But I guess both of these have so few functions that it's not critical to split the section.

**Comment by szhorvat** [comment]
Should `console()` go under "Interactive functions"? I've never used it. But it does not work at all!

**Comment by ntamas** [comment]
> I don't like this very much. Do you agree that we should keep `ecount()` and `vcount()` as fully supported names?

It's fine for me if we keep them, I don't have any preference. Feel free to remove any mentions of them being deprecated if you want to.

> "nexus" functions are not really deprecated?

They aren't in the formal sense, but I highly doubt that anyone used them in the last five years or so because the server was gone ages ago.

> I'd recommend exploring https://lifecycle.r-lib.org/

That's a very good reference, thanks! I'll take a look at it.

**Comment by maelle** [comment]
(preview still building, I'll update the PR once it's up)

- [x] Can you put a single function in more than one section? [simplify_and_colorize()](https://maelle.github.io/rigraph/reference/simplify.html) can be placed under isomorphism. I implemented this function specifically for dealing with isomorphism testing of non-simple graphs.

Yes one can :tada: 

- [x] If a function can appear in multiple sections, (make_)from_prufer should also appear under Trees
- [x] The "Simple" section can be merged with "Functions for manipulating graphs"
- [x] I see that now there are subsections. "Random graph models" should be a subsection of Construction then.
- [x] Do not separate isomorphism from automorphisms. They are the same thing. Do not use the title "Topology" (which is mathematically inaccurate), just use "Isomoprhism".
- [x] Rename Flow to "Maximum flow and connectivity"
- [x] Centalization should come right after Centrality measures, as it's closely related.

However centrality is a subsection of the previous section so they're not really together, is that a problem?

- [x]  triangles() is missing for some reason.

The reason was a roxygen2 tag problem, apparently.

- [x] Under Structural properties, let's make a subsection on "Triangles and transitivity". This would include all functions with "triangle" or "transitivity" in their names.

Luckily there's even a pkgdown reference for that, `matches()`! 

- [x] The "Graph matching" subsection of Structural properties should be moved to the top level (or possibly to a subsection of Isomorphism).

- [x] "Processes on graphs" could have two subsections, "Compartmental models" for SIR and "Random walks". But I guess both of these have so few functions that it's not critical to split the section.

So not splitting for now, is that fine?

**Comment by maelle** [comment]
https://maelle.github.io/rigraph/reference/index.html

**Comment by szhorvat** [comment]
IMO this looks really nice and most of the remaining polishing can be done after merging the PR.

Some things that stand out for me:

 - If I click a section name, subsection names won't immediately show in the sidebar. They only show after I scroll down a bit. I don't know if this is fixable.
 - "Random graph models" should follow "Deterministic constructors" directly. Perhaps some of you prefer "Stochastic constructors" instead of "Random graph models"?
 - Regarding "I/O Read/Write Files", let's keep the capitalization consistent with the rest
 - Earlier I suggested renaming "Functions for manipulating graph structure" to "Functions for manipulating graphs" to make it a bit broader. Just want to check whether you left the original title on purpose, or missed my comment. If you prefer that original title, that's fine.
 - Is it easy to sort functions by name under each (sub)section? E.g. the contents of "Functions for manipulating graph structure" is _mostly_ sorted, but it also has a few extra entries at the end. At first I thought it did not include `permute()` as it's not in sorted order.
 - There are some deprecated layout functions in the "Versions" section. These don't belong there. I would actually suggest removing these completely from the docs, as people shouldn't be using them (@ntamas, are you okay with that?)
 - `srand()` is also a deprecated (and non-working) function that should not be in the docs. 
 - IMO `erdos.renyi.game()` should not be included. We have the consistently named `sample_gnm` and `sample_gnp` that provide the exact same functionality. I am not sure that everyone agrees with me on this point, so ping @ntamas again.
 - IMO `graph_` and `graph_from_lcf()` should go under "Deterministic constructors", just like other `graph_from_...` functions. My reasoning for this is that we should only call something "conversion" if it can be bidirectional in a practical sense. With most of these, it is not. Conversion _to_ a literal (`graph_from_literal()`) is technically possible but does not make sense. Converting to LCF notation seems difficult (and only possible for a small subset of graphs). For `graph_from_atlas()` back-conversion makes no sense.
 - Now that I understand that we can have the same function in multiple sections, I do suggest including `permute()` in Isomorphism as well. 
 - @ntamas Does `layout.grid.3d` count as deprecated or not? I don't like that it uses the old naming. We have the alternative `layout_on_grid(..., dim=3)`. But I don't like that one must specify `dim=3` as a parameter in `layout_on_grid` instead of having a dedicated function. OTOH this is how `layout_with_fr` and `layout_with_kk` work too, so it's consistent. I'd suggest excluding `layout.grid.3d` from the reference page.
 - Sorry about contradicting myself from earlier: I would actually make "Graph matching" a top-level section that follows "Graph isomorphism" immediately. It looks a bit weird as it is at the moment.
 - We have a "Convert to igraph" subsection in "Constructors" as well as a "Conversion" subsection in "Utilities". Do we need both? Should the former be merged into the latter? This is for someone with R experience to decide, so make your pick @maelle 
 - Can we expand "Misc" to "Miscellaneous"?
 - "Latent position vector samplers" are a niche things that should probably not be at the top level. These are meant to be used with `sample_dot_product()`. Is there a way to tie them to that function? Otherwise do people think this is better as a subsection under "Utilities"?
 - Centalization should come right after Centrality measures, as it's closely related. I suggest renaming "Centralization related" to simply "Centralization".

**Comment by szhorvat** [comment]
I would also suggest a reordering of sections. This is the current ordering:

```
    igraph options
    Construction
    Visualization
    Functions for manipulating graph structure
    Utilities
    Structural properties
    Centralization related
    Graph motifs and subgraphs
    Graph isomorphism
    Maximum flow and connectivity
    Cliques
    Community detection
    Graph cycles
    Processes on graphs
    Vertex, edge and graph attributes
    Cliques
    Connected components
    Graphical degree sequences
    Demo
    Latent position vector samplers
    I/O Read/Write Files
    Hierarchical random graph functions
    Vertex and edge sequences
    Functions for manipulating graph structure
    Scan statistics
    Spectral Coarse Graining
    Embedding
    Interactive functions
    Versions
```

Here I propose a reordering. I mark those sections which I moved by a `*`, and try to explain the reasoning behind the reordering.

```
    igraph options
    Construction
    Visualization
    Functions for manipulating graph structure
    Functions for manipulating graph structure * (appears 2nd time, this should be fixed)
    Vertex, edge and graph attributes *
    Vertex and edge sequences *
    Utilities
----< network analysis start
    Structural properties
    Centralization related <- move into a subsection of Structural properties
    Graph motifs and subgraphs
    Graph isomorphism
    Maximum flow and connectivity
    Cliques
    Community detection
    Graph cycles
    Cliques
    Connected components
    Scan statistics
    Spectral Coarse Graining
    Embedding
    Hierarchical random graph functions
----> network analysis end, we keep these sections together

    Graphical degree sequences
    Processes on graphs

    Demo
    Latent position vector samplers <- move into subsection of Utilities
    I/O Read/Write Files
    Interactive functions
    Versions
```

**Comment by szhorvat** [comment]
Final comment: should the Construction subsection be moved from Utilities into Construction?

**Comment by ntamas** [comment]
It's shaping up nicely; I'm happy to merge this PR whenever you guys consider it ready.

**Comment by ntamas** [comment]
Actually, how about merging it now as it's already a big improvement over what we have now? And then we can keep on iterating after that in a separate PR.

**Comment by szhorvat** [comment]
Do you have any comments on the points where I pinged you specifically? Let me know if you disagree with any of my suggestions there.

As for merging, that's an organizational question, up to what's most convenient for @maelle. No reason not to merge now.

**Comment by ntamas** [comment]
> There are some deprecated layout functions in the "Versions" section. These don't belong there. I would actually suggest removing these completely from the docs, as people shouldn't be using them

I'm okay with leaving these functions out of the table of contents as long as they are still discoverable somehow in case someone needs them. (Like, Google should be indexing the documentation of these functions even if they are deprecated in case someone searches Google for them). The same applies for `erdos.renyi.game`, i.e. we now have `sample_gnm` and `sample_gnp`, and these are better, but still some people might look for the older name in Google.

> Does `layout.grid.3d` count as deprecated or not? I don't like that it uses the old naming.

I think that practically all "dotted" names for which we have a newer, underscore-style name are deprecated. Strangely enough, some of the old "dotted" names do not have alternatives yet, which might be an oversight. I don't really know enough about the history of R to know why people are moving away from "dotted" names, but if it's a trend and igraph would feel more "native" in contemporary R with underscored names, then we should probably provide alternatives for all of the old "dotted" names.

**Comment by ntamas** [comment]
@krlmlr @maelle Do you know why CI is failing for this PR consistently? I see no reason why it fails here when the same pipeline works in `main`.

**Comment by szhorvat** [comment]
@ntamas If you're not comfortable with leaving these out of the docs, we could have a "Deprecated" section for them, like we do in the C core.

**Comment by szhorvat** [comment]
This will replace the page at https://igraph.org/r/doc/, right?  Will there still be a way to look at the docs of older versions?

image

If yes, that would alleviate the problem you mentioned, i.e. that people find some of the deprecated functions using google and want to read the documentation for them.

**Comment by ntamas** [comment]
> Will there still be a way to look at the docs of older versions?

idk, what's the pkgdown way of doing this? (if they do this in the R world at all...)

> we could have a "Deprecated" section for them

that's fine by me

**Comment by maelle** [comment]
Answer regarding version selector, it's not implemented in pkgdown yet. https://github.com/r-lib/pkgdown/pull/2072

**Comment by maelle** [comment]
@krlmlr regarding the failure of R-CMD-check, could you have a look at the log / give me any tip as it's related to the Make file I haven't explored at all yet?

**Comment by maelle** [comment]
> I'm okay with leaving these functions out of the table of contents as long as they are still discoverable somehow in case someone needs them. (Like, Google should be indexing the documentation of these functions even if they are deprecated in case someone searches Google for them). 

They'll also be index in pkgdown search index (for the search bar in the navbar).

Generally {lifecycle} would help with badging / messaging around deprecation.

For inspiration, tidyr has a reference section for _superseded_ function which if I follow correctly is the case for some of the functions you mention https://tidyr.tidyverse.org/reference/index.html#superseded

**Comment by maelle** [comment]
I'll get to https://github.com/igraph/rigraph/pull/662#issuecomment-1419425495 later this week or next.

**Comment by maelle** [comment]
(let's not merge before all checks are green, as I might have inadvertently broken the build :cold_sweat:)

**Comment by krlmlr** [comment]
The failure is triggered by `man/roxygen/meta.R` . Do we need it in this PR, can we extract it to a separate PR?

**Comment by maelle** [comment]
> Do we need it in this PR, can we extract it to a separate PR?

We can indeed, it just improves the Rd. I will tackle this with the rest of the comments some time this or next week.

**Comment by szhorvat** [comment]
There are some tutorial pages in the documentation. See e.g. `?igraph.plotting`. Can we have a table of contents for these, so they are more discoverable?

**Comment by maelle** [comment]
@szhorvat 

> Final comment: should the Conversion subsection be moved from Utilities into Construction?

The functions in that subsection do not convert _to_ graphs, at least of them convert _from_ graphs. Do you still want to move them into construction?

**Comment by maelle** [comment]
- [ ]  If I click a section name, subsection names won't immediately show in the sidebar. They only show after I scroll down a bit. I don't know if this is fixable.

This is either a pkgdown or bootstrap-toc bug(let)/unexpected behavior.
On https://afeld.github.io/bootstrap-toc/ if I scroll down then click on usage, I also don't see the subsections before scrolling down. However if I click on "Customization" I see the subsections.

- [x] "Random graph models" should follow "Deterministic constructors" directly. Perhaps some of you prefer "Stochastic constructors" instead of "Random graph models"?
- [x] Regarding "I/O Read/Write Files", let's keep the capitalization consistent with the rest
- [x] Earlier I suggested renaming "Functions for manipulating graph structure" to "Functions for manipulating graphs" to make it a bit broader. Just want to check whether you left the original title on purpose, or missed my comment. If you prefer that original title, that's fine.
- [x] Is it easy to sort functions by name under each (sub)section? E.g. the contents of "Functions for manipulating graph structure" is mostly sorted, but it also has a few extra entries at the end. At first I thought it did not include permute() as it's not in sorted order.

It was because they didn't all belong to the same family, I added a common family tag to all of them now.

- [ ] There are some deprecated layout functions in the "Versions" section. These don't belong there. I would actually suggest removing these completely from the docs, as people shouldn't be using them (@ntamas, are you okay with that?)

Which ones?

- [x] srand() is also a deprecated (and non-working) function that should not be in the docs.

I added keywords internal to it but I think consolidating the lifecycle workflow (using the lifecycle package?) would make sense.

- [ ] IMO erdos.renyi.game() should not be included. We have the consistently named sample_gnm and sample_gnp that provide the exact same functionality. I am not sure that everyone agrees with me on this point, so ping @ntamas again.

Same comment as above.

- [x] IMO graph_ and graph_from_lcf() should go under "Deterministic constructors", just like other graph_from_... functions. My reasoning for this is that we should only call something "conversion" if it can be bidirectional in a practical sense. With most of these, it is not. Conversion to a literal (graph_from_literal()) is technically possible but does not make sense. Converting to LCF notation seems difficult (and only possible for a small subset of graphs). For graph_from_atlas() back-conversion makes no sense.
- [x] Now that I understand that we can have the same function in multiple sections, I do suggest including permute() in Isomorphism as well.
- [ ] @ntamas Does layout.grid.3d count as deprecated or not? I don't like that it uses the old naming. We have the alternative layout_on_grid(..., dim=3). But I don't like that one must specify dim=3 as a parameter in layout_on_grid instead of having a dedicated function. OTOH this is how layout_with_fr and layout_with_kk work too, so it's consistent. I'd suggest excluding layout.grid.3d from the reference page.
- [x] Sorry about contradicting myself from earlier: I would actually make "Graph matching" a top-level section that follows "Graph isomorphism" immediately. It looks a bit weird as it is at the moment.
- [ ] We have a "Convert to igraph" subsection in "Constructors" as well as a "Conversion" subsection in "Utilities". Do we need both? Should the former be merged into the latter? This is for someone with R experience to decide, so make your pick @maelle

I'm not sure as the conversion functions do not construct a graph.

- [x] Can we expand "Misc" to "Miscellaneous"?
- [x] "Latent position vector samplers" are a niche things that should probably not be at the top level. These are meant to be used with sample_dot_product(). Is there a way to tie them to that function? Otherwise do people think this is better as a subsection under "Utilities"?
- [x] "Centralization" should come right after "Centrality measures", as a subsection, as it's closely related. I suggest renaming "Centralization related" to simply "Centralization".
- [ ] Should "Scan statistics" be a subsection of "Structural properties"? Thoughts?

Who should decide?

**Comment by ntamas** [comment]
> There are some deprecated layout functions in the "Versions" section. These don't belong there.

I don't see them either; I guess this point is now obsolete.

> I think consolidating the lifecycle workflow (using the lifecycle package?) would make sense.

Totally agree, but I think that is not in the scope of this PR; we could do it separately in another PR. If you think that it's easier to make the deprecated functions disappear using the `lifecycle` package, just leave them as is for now and we'll update the docs when we adopt the usage of the `lifecycle` package.

> Does layout.grid.3d count as deprecated or not?

Yes, it's deprecated.

> Should "Scan statistics" be a subsection of "Structural properties"? Thoughts?

Let's make it a subsection.

**Comment by szhorvat** [comment]
`console()` can go under "Interactive", it's fixed now.

> * There are some deprecated layout functions in the "Versions" section. These don't belong there. I would actually suggest removing these completely from the docs, as people shouldn't be using them (@ntamas, are you okay with that?)
> 
> 
> Which ones?

I guess these are not really under "Versions", they just come at the end and _appear_ to be under "Versions"? See here:

https://maelle.github.io/rigraph/reference/index.html#versions

I think everything should be excluded, _except_:

 - `graph_version()`
 - `upgrade_graph()`
 - `igraph_version()`
 - `igraph_test()`
 - `console()` (move to "Inrteractive")
 - `%>%` (but do we need this in the igraph docs?)

If this section is renamed to something else, e.g. "System", then `igraph_test()` will fit here just fine. Otherwise it needs a new home.

**Comment by maelle** [comment]
I had make a mistake in the pkgdown config syntax so all functions I had listed under "Internal" still showed up, sorry about that.

Are there further tweaks to make?

**Comment by maelle** [comment]
@krlmlr apart from the conflicts, ok to merge this?

**Comment by szhorvat** [comment]
Can we rename "Stochastic constructors – Random graph models (games)" to simply "Stochastic constructors (random graph models)"?  I know I suggested the current title, but I just realize that the new dot-free names no longer use the "game"-terminology. So there's no reason to put that into the title.

**Comment by ntamas** [comment]
@maelle This conflicts with the base branch - is there an easy way to resolve the conflicts so this can be merged?

**Comment by maelle** [comment]
> @maelle This conflicts with the base branch - is there an easy way to resolve the conflicts so this can be merged?

@ntamas I don't think they are huge conflicts but the interface tells me that only those with write access can merge the PR?

**Comment by krlmlr** [comment]
I resolved the conflicts by merging main into this branch and rerunning `devtools::document()` .

@maelle: Would you like to do one more iteration here?

**Comment by maelle** [comment]
I changed the default branch which was causing quite a few problems when trying to fix conflicts.

Rebasing didn't work as I expected (I'm stuck a conflict in cigraph/), so I'll try merging as indicated above.

**Comment by maelle** [comment]
@ntamas @szhorvat time to merge or time for a few last tweaks? :smile_cat:

File: .github/workflows/pkgdown.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@
 # Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help
 on:
   push:
-    branches: [main, master, config-draft]
+    branches: [main, master, reference-reorg]
```
Comment (szhorvat) [inline .github/workflows/pkgdown.yaml:5]: Is a mention of any branch other than `main` still needed after this PR is merged?

Comment (maelle) [reply .github/workflows/pkgdown.yaml]: Good catch!

**Pull request merged**